pyboolnet/control_strategies.py

```diff
@@ -0,0 +1,486 @@
+
+
+def is_control_strategy(primes: dict, candidate: dict, target: List[dict], update: str, max_output: int = 1000000):
+    """
+    Check whether the *candidate* subspace is a control strategy for the *target* subset.
+
+    **arguments**:
+        * *primes*: prime implicants.
+        * *candidate*: candidate subspace.
+        * *target*: list of subspaces defining the target subset.
+        * *update*: type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+        * *max_output*: maximum number of trap spaces computed. Default value: 1000000.
+
+    **returns**:
+        * *answer*: True if the *candidate* is a control strategy for the *target*, False otherwise. When False, a counterexample (minimal trap space disjoint with target or state not satisfying the CTL query) is also returned.
+
+    **example**::
+        >>> is_control_strategy(primes, {'v1': 1}, [{'v2': 0}, {'v3':1}], "asynchronous")
+    """
+
+    # Preliminary setting
+
+    if type(target) == dict:
+        print("The target must be a list of subspaces.")
+        return
+
+    if control_direct_percolation(primes, candidate, target, silent=True):
+        return True
+
+    new_primes = fix_components_and_reduce(primes, candidate, list(set([x for subs in subspaces for x in subs])))
+    tsmin = compute_trap_spaces(new_primes, "min", max_output=max_output)
+
+    # Checking control is valid in the trap spaces
+    if not control_is_valid_in_trap_spaces(new_primes, tsmin, target, update):
+        return False
+
+    # Checking general CTL query
+    return run_control_query(new_primes, target, update)
+
+
+def compute_control_strategies_with_completeness(primes: dict, target: dict, update: str = "asynchronous", limit: int = 3, avoid_nodes: List[str] = None, silent: bool = False, starting_length: int = 0, previous_cs: List[dict] = None, known_cs: List[dict] = None) -> List[dict]:
+    """
+    Identify control strategies for the *target* subspace using completeness.
+
+    **arguments**:
+        *primes*: prime implicants.
+        *target*: target subspace.
+        *update*: the type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+        *limit*: maximal size of the control strategies. Default value: 3.
+        *silent*: if True, it does not print infos to screen. Default value: False.
+        *starting_length*: minimum possible size of the control strategies. Default value: 0.
+        *previous_cs*: list of already identified control strategies. Default value: empty list.
+        *avoid_nodes*: list of nodes that cannot be part of any control strategy. Default value: empty list.
+
+    **returns**:
+        * *cs_total*: list of control strategies for the *target* subspace obtained using completeness.
+
+    **example**::
+        >>> control_strategies_completeness(primes, {'v1': 1}, "asynchronous")
+
+    """
+
+    # Initializing
+
+    if previous_cs is None:
```

Comment:
you can use

```
previous_cs = previous_cs or []
```

please don't use abbreviations (use previous_controls or previous_strategies or ..)